File: src/FsApi/Controllers/Store.fs
namespace Api.Controllers

open Microsoft.AspNetCore.Mvc
open Microsoft.Extensions.Logging
open Api
open Microsoft.AspNetCore.Authorization
open AspNetExtensions
open Microsoft.AspNetCore.Http
open System.Threading.Tasks


module LocalStore =
    let private ensureStoreExists uri =
        let storePath = uri |> FS.fullPath
        match uri with
        | FS.File _ -> failwith "Found file {uri} instead of directory"
        | FS.Directory _ -> ()
        | _ -> uri |> IO.createDirectory
        storePath

    let exists uri path =
        let storePath = ensureStoreExists uri
        let artifactPath = FS.combinePath storePath path |> FS.fullPath
        printfn $"Local: exists {artifactPath}"
        
        // security note: we must ensure path within store path
        if artifactPath |> String.startsWith storePath |> not then false
        else
            match artifactPath with
            | FS.File _ -> true
            | _ -> false

    let read uri path =
        let storePath = ensureStoreExists uri
        Array.empty<byte>

    let write uri path =
        let storePath = ensureStoreExists uri
        false

[<ApiController; Route("store")>]
type StoreController (logger : ILogger<AuthController>, appSettings: AppSettings) =
    inherit ControllerBase()

    [<HttpGet; Route("exists")>]
    member _.Exists([<FromQuery>] path: string): ActionResult =
        printfn $"Exists {path}"

        match appSettings.Store.Type with
        | StoreType.Local -> if LocalStore.exists appSettings.Store.Uri path then base.Ok() else base.NotFound()
        | _ -> failwith $"Unknown store type {appSettings.Store.Type}"

    [<HttpGet; Route("read/{path}")>]
    member _.Read([<FromQuery>] path: string): Task<byte array> =
        match appSettings.Store.Type with
        | StoreType.Local -> LocalStore.read appSettings.Store.Uri path |> Task.FromResult
        | _ -> failwith $"Unknown store type {appSettings.Store.Type}"

    [<HttpPut; Route("write/{path}")>]
    member _.Write(path: string): ActionResult =
        match appSettings.Store.Type with
        | StoreType.Local -> if LocalStore.write appSettings.Store.Uri path then base.Ok() else base.NotFound()
        | _ -> failwith $"Unknown store type {appSettings.Store.Type}"

File: src/CsApi/Controllers/Store.cs
namespace Api.Controllers;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Logging;
using Api;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Http;
using System.Threading.Tasks;
using System.IO;
using Microsoft.AspNetCore.Http.HttpResults;
using Microsoft.AspNetCore.Razor.TagHelpers;

public static class LocalStore {
    public static string EnsureStoreExists(string uri) {
        var storePath = Path.GetFullPath(uri);
        if (!Directory.Exists(storePath)) {
            Directory.CreateDirectory(storePath);
        }

        return storePath;
    }


    public static bool Exists(string uri, string path) {
        var storePath = EnsureStoreExists(uri);
        var artifactPath = Path.GetFullPath(Path.Combine(storePath, path));
        if (!artifactPath.StartsWith(storePath)) {
            return false;
        }

        return File.Exists(artifactPath);
    }

    public static byte[] Read(string uri, string path) {
        var storePath = EnsureStoreExists(uri);
        return [];
    }

    public static bool Write(string uri, string path) {
        return false;
    }
}

[ApiController, Route("store")]
public class StoreController(ILogger<AuthController> logger, AppSettings appSettings) : ControllerBase {

    [HttpGet, Route("exists")]
    public ActionResult Exists([FromQuery] string path) {
        switch (appSettings.Store.Type) {
            case StoreType.Local:
                if (LocalStore.Exists(appSettings.Store.Uri, path)) {
                    return Ok();
                } else {
                    return NotFound();
                }

            default:
                return BadRequest();
        }
    }

    [HttpGet, Route("read")]
    public async Task<ActionResult<byte[]>> Read([FromQuery] string path) {
        switch (appSettings.Store.Type) {
            case StoreType.Local:
                return LocalStore.Read(appSettings.Store.Uri, path);

            default:
                return BadRequest();
        }
    }
}
